urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/httpd/CustomUtil.java

```diff
@@ -0,0 +1,536 @@
+	    }
+	}
+	
+	return arr;
+    }
+    
+    public static byte[] hex2bytes( String hex ) 
+	throws NumberFormatException {
+
+	if( hex == null )
+	    return null;
+
+	if( hex.length()%2 != 0 )
+	    throw new NumberFormatException( "Cannot convert hex strings with odd numbers of characters to byte arrays." );
+
+	byte[] arr = new byte[ hex.length() / 2 ];
+
+	int hi, lo;
+	char c_hi, c_lo;
+	for( int i = 0; i < arr.length; i++ ) {
+
+	    c_hi = hex.charAt( i*2 );
+	    c_lo = hex.charAt( i*2 + 1);
+	    
+	    hi = hex2int( c_hi );
+	    lo = hex2int( c_lo );
+	    //System.out.println( "hi=" + c_hi + "(" + hi + "), lo=" +  c_lo + "(" + lo + ")" );
+
+	    arr[i] = (byte)((hi << 4) | lo);
+
+	    
+	}
+
+	return arr;
+    }
+
+ 
+    public static int hex2int( char hex ) 
+	throws NumberFormatException {
+
+	// '0' to '9' in ASCII: 48 to 57 (dec)
+	// 'A' to 'F' in ASCII: 65 to 70 (dec)
+	// 'a' to 'f' in ASCII: 97 to 102 (dec)
+	
+	if( hex >= 48 && hex <= 57 ) {
+
+	    // '0' to '9'
+	    return (hex-48);
+
+	} else if( hex >= 97 && hex <= 102 ) {
+
+	    // 'a' to 'f'
+	    return (10 + hex-97);
+
+	} else if( hex >= 65 && hex <= 70 ) {
+
+	    // 'A' to 'F'
+	    return (10 + hex-65);
+
+	} else {
+
+	    throw new NumberFormatException( "Cannot convert char '"+hex+"' to hex-nibble (not a hexadecimal character)." );
+
+	}
+
+    }
+
+    public static String bytes2hexString( byte[] arr ) {
+	
+	StringBuffer b = new StringBuffer();
+
+	for( int i = 0; i < arr.length; i++ ) {
+
+	    b.append( upperNibble2hexChar(arr[i]) );
+	    b.append( lowerNibble2hexChar(arr[i]) );
+
+	}
+
+	return b.toString();
+    }
+
+    public static char lowerNibble2hexChar( byte b ) {
+	return Integer.toHexString( (b & 0x0F) ).charAt(0);
+    }
+
+    public static char upperNibble2hexChar( byte b ) {
+	return Integer.toHexString( ((b>>4) & 0x0F) ).charAt(0);
+    }
+    
+
+    public static long transfer( InputStream in,
+				 OutputStream out,
+				 long maxReadLength,
+				 int bufferSize ) 
+	throws IllegalArgumentException,
+	       IOException {
+
+	byte[] buffer = new byte[ bufferSize ];
+	int len;
+	long totalLength = 0;
+	long bytesLeft = maxReadLength; // Math.max( maxReadLength, bufferSize );
+
+	while( (maxReadLength == -1 || totalLength < maxReadLength)
+	       //&& (len = in.read(buffer,0,(int)bytesLeft)) > 0 ) {
+	       && (len = in.read(buffer,0,(int)Math.max( maxReadLength, bufferSize ))) != -1 ) {
+
+	    //for( int i = 0; i < len; i++ )
+	    //	System.out.print( "X" + (char)buffer[i] );
+
+	    out.write( buffer, 0, len );
+	    out.flush();
+	    
+	    totalLength += len;
+	    bytesLeft -= len;
+	}
+
+	return totalLength;
+    }
+
+    // For testing only
+    public static void main( String[] argv ) {
+
+	String challenge = "username=\"keel\", realm=\"Password Required\", nonce=\"123456\", uri=\"/secret\", algorithm=MD5, response=\"c21340f3fdd7194ab8e397ff1a9a750f\"";
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It appears your code contains a hardcoded Shopify Private App Password. Hardcoded secrets or credentials can allow attackers to bypass authentication methods and perform malicious actions. We recommend revoking access to resources using this credential and storing future credentials in a management service such as [AWS Secrets Manager](https://aws.amazon.com/secrets-manager/).

[Learn more about the use of hardcoded credentials](https://cwe.mitre.org/data/definitions/798.html)